### PR TITLE
Moving em_progenitors.py and ns_sequence data from tmpltbank to new neutron_stars package

### PR DESCRIPTION
**Pull request by SamuelH-97**
The functions within tmpltbank/em_progenitors.py serve purposes beyond template banks, for instance, they are used quite integrally in pygrb where the remnant mass function is used to separate potentially EM-bright injections from EM-dim injections. Here, I have moved the remnant_mass function to conversions.py so it may be used more broadly, and moved the rest of the em_progenitors.py functions to the new ns_functions.py (name subject to change) within a new package under the name neutron_stars. This package also now houses the ns_sequence data once stored in tmpltbank since it is more relevant to the functions stored here. The aim of this change is to allow for these functions to be used more broadly under a more appropriately named package.

### CODE REVIEW

File: pycbc/tmpltbank/coord_utils.py [hunk 1]
```diff
@@ -20,7 +20,8 @@
 from pycbc.tmpltbank.lambda_mapping import get_chirp_params
 from pycbc import conversions
 from pycbc import pnutils
-from pycbc.tmpltbank.em_progenitors import load_ns_sequence, remnant_masses
+from pycbc.neutron_stars import load_ns_sequence
+from pycbc.conversions import remnant_mass
```
Comment (a-r-williamson) [inline pycbc/tmpltbank/coord_utils.py:24]:
```suggestion
```

File: pycbc/tmpltbank/coord_utils.py [hunk 2]
```diff
@@ -292,7 +293,7 @@ def get_random_mass(numPoints, massRangeParams):
             #     threshold mass]
             mask_bright_nsbh = numpy.zeros(len(mass1_nsbh), dtype=bool)
             if eta_nsbh.size != 0:
-                remnant = remnant_masses(eta_nsbh, mass2_nsbh, ns_sequence,
+                remnant = remnant_mass(eta_nsbh, mass2_nsbh, ns_sequence,
                                          spin1z_nsbh, 0.)
```
Comment (a-r-williamson) [inline pycbc/tmpltbank/coord_utils.py:297]:
```suggestion
                remnant = conversions.remnant_mass(eta_nsbh, mass2_nsbh, ns_sequence,
                                                   spin1z_nsbh, 0.)
```

File: pycbc/conversions.py [hunk 3]
```diff
@@ -1545,3 +1550,82 @@ def nltides_gw_phase_diff_isco(f_low, f0, amplitude, n, m1, m2):
...
+    remnant_mass = remnant_mass*ns_b_mass
+
+    return remnant_mass
+
+
+#############################################################################
+# Vectorized version of remnant_mass. The third argument (NS equilibrium    #
+# sequence) is excluded from vectorisation.                                 #
+#############################################################################
+remnant_masses = np.vectorize(remnant_mass)
```
Comment (a-r-williamson) [inline pycbc/conversions.py:1630]:
```suggestion
remnant_masses = numpy.vectorize(remnant_mass)
```

File: pycbc/neutron_stars/ns_functions.py [hunk 4]
```diff
@@ -174,11 +174,12 @@ def PG_ISSO_solver(chi,incl):
...
     return solution
 
+
 ##############################################################################
 # 2H 2-piecewise polytropic EOS, NS non-rotating equilibrium sequence        #
 # File format is: grav mass (Msun)   baryonic mass (Msun)    compactness     #
-#                                                                            #
-# Eventually, the function should call an NS sequence generator within LAL   #
+#                                                                            #       THIS FUNCTION IS BEING CALLED IN TMPLTBANK, does it need to be moved back?
+# Eventually, the function should call an NS sequence generator within LAL   #       or should I go and modify those files to reflect new location?
```
Comment (a-r-williamson) [inline pycbc/neutron_stars/ns_functions.py:182]: Anywhere these functions are called from should be updated to point to the new location (and then remove this comment)

Comment (SamuelH-97) [reply pycbc/neutron_stars/ns_functions.py:182]: Sorry about this, that was a very old comment I didn't spot before the PR, just fixed it.

File: pycbc/conversions.py (hunk 3 above)
Comment (SamuelH-97) [reply pycbc/conversions.py:1630]: I'm asking this with regards to the other suggested changes above, are these just to trim down unnecessary imports? Does this affect performance in any way or is it mainly a style preference for pycbc?

Comment (SamuelH-97) [reply pycbc/conversions.py:1630]: Actually I notice the issue with the coord_utils change. Completely missed the double import there, thanks for pointing that out.

File: pycbc/conversions.py [hunk 5]
```diff
@@ -1545,3 +1549,82 @@ def nltides_gw_phase_diff_isco(f_low, f0, amplitude, n, m1, m2):
...
+    if not (eta>0. and eta<=0.25 and abs(chi)<=1 and ns_g_mass>0):
+        err_msg = "The BH spin magnitude must be <= 1,"
+        err_msg += "eta must be between 0 and 0.25,"
+        err_msg += "and the NS mass must be positive."
+        logging.error(err_msg)
+        info_msg = "The function remnant_mass was launched with"
+        info_msg += "ns_mass={0}, eta={1}, chi={2},"
+        info_msg += "inclination={3}\n'.format(ns_g_mass, eta, chi, incl)"
+        logging.info(info_msg)
+        sys.exit(1)
```
Comment (titodalcanton) [inline pycbc/conversions.py:1603]: Please remove the `logging.error()` and `sys.exit()` calls, change the exception to a `ValueError`, and show the full error message in the exception instead.

Comment (titodalcanton) [reply pycbc/conversions.py:1603]: Then you can also remove the sys import above.

File: pycbc/conversions.py [hunk 6]
```diff
@@ -1545,3 +1549,82 @@ def nltides_gw_phase_diff_isco(f_low, f0, amplitude, n, m1, m2):
...
+        raise Exception('Unphysical parameters!')
```
[the remark this answers is not in the document]
Comment (a-r-williamson) [reply pycbc/conversions.py:1604]: Suggest this solution here, then remove `logging` and `sys` imports.

File: pycbc/conversions.py [hunk 7]
```diff
@@ -32,10 +32,14 @@
 import copy
 import numpy
 import lal
+import logging
+import sys
```
Comment (a-r-williamson) [inline pycbc/conversions.py:36]:
```suggestion
```

File: pycbc/conversions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (SamuelH-97) [reply pycbc/conversions.py:1604]: Thanks for the suggested changes

**Comment by rpfisher** [comment]
@a-r-williamson @SamuelH-97 @pannarale @jakeb245 @nathan.ormsby 
Hi Samuel (and possibly Andrew).  Jacob Buchanan and Nathan Ormsby would like to help with the injection part of the PyGRB workflow if possible. We've taken a look through this PR and the linked PR (not all the code in detail), and would like to know if there is a specific issue or problem we could tackle.  It looks like the major problem/issue is editing the code to make it mergeable with the upstream branch?  If you can just give us one thing to do, we'll start diving into the code and take it on!

**Comment by titodalcanton** [comment]
Starting to look at this again. It seems codeclimate has picked up a few genuine issues, for example there is a reference to an undefined variable, and some docstrings have backslashes which are being interpreted as escape sequences. Could someone take a look at those things?

**Comment by SamuelH-97** [comment]
> Starting to look at this again. It seems codeclimate has picked up a few genuine issues, for example there is a reference to an undefined variable, and some docstrings have backslashes which are being interpreted as escape sequences. Could someone take a look at those things?

Thanks for the heads up. The undefined variable was an error on my part in a recent change and has bee resolved. 

As for the backslashes, they seem to be a part of the equation formatting for automated documentation. I'm unsure if he has the time to check these but perhaps @a-r-williamson could provide some wisdom here. I'm somewhat unfamiliar with these changes to the docstrings and how they may affect both the everyday running of the code and the documentation

**Comment by spxiwh** [comment]
@SamuelH-97 For the docstrings I think you just need to add the "r" prefix: https://developer.lsst.io/python/numpydoc.html#docstrings-must-be-delimited-by-triple-double-quotes

which should not mess up the latex. In this case I think it's okay as there's nothing special in here, but try and use `\tan` and you'll probably spot the problem!!

File: pycbc/neutron_stars/eos_utils.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        Name of the equation of state.
+
+    Returns
+    -------
+    ns_compactness : float
+        Compactness parameter of the neutron star.
+    ns_b_mass : float
+        Baryonic mass of the neutron star.
+    """
+    if str(type(ns_mass)) == '<class \'numpy.ndarray\'>':
```
Comment (titodalcanton) [inline pycbc/neutron_stars/eos_utils.py:132]: I think we usually want to do this kind of check like this:
```suggestion
    if isinstance(ns_mass, numpy.ndarray):
```
Also, update the docstring above to say that `ns_mass` can also be a Numpy array, not just a float.

File: pycbc/neutron_stars/eos_utils.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        Baryonic mass of the NS.
+    bh_spin_mag: float
+        Dimensionless spin magnitude of the BH.
+    bh_spin_pol : float
+        The tilt angle of the BH spin.
+    interp: bool, optional
+        Whether to interpolate over a pre-computed grid of values of
+        the ISSO for systems with BH spin magnitude < 0.9. Otherwise
+        solve via the PG_ISSO_solver routine for all systems.
+        Default = True.
```
Comment (titodalcanton) [inline pycbc/neutron_stars/eos_utils.py:189]: Check this, it looks like some of those parameters can take Numpy arrays, not just float.

File: pycbc/neutron_stars/pg_isso_solver.py [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+        (0.125 * np.pi, 50), (0.375 * np.pi, 300), (0.5 * np.pi, 50),
+        (0.9 * np.pi, 50), (np.pi, 26)))
+    chig, incg = np.meshgrid(chis, incs)
+    roots = np.empty_like(chig)
+    for i, (chi, inc) in enumerate(zip(chig, incg)):
+        # more dense at larger chi
+        roots[i] = PG_ISSO_solver(chi, inc)
+    bivar = RectBivariateSpline(incs, chis, roots)
+    with open(os.path.join(NS_DATA_DIRECTORY, 'isso_inc_chi.pkl'), 'wb') as f:
+        pickle.dump(bivar, f)
```
Comment (titodalcanton) [inline pycbc/neutron_stars/pg_isso_solver.py:423]: This looks like it is trying to write to a file in a directory within the PyCBC install path, which is generally a horrible idea, and more importantly, will not work in general. For example, it will definitely fail in a IGWN Conda install of PyCBC. It is not clear what the solution is here, maybe the function should take the output file path?

Comment (titodalcanton) [reply pycbc/neutron_stars/pg_isso_solver.py:423]: Also, a file with the name `isso_inc_chi.pkl` is being committed by this PR. Was that file generated by calling this function?

**Comment by titodalcanton** [comment]
I took another deep look through, and left some trivial comments and a more substantial one. I think those will be my last comments.

File: pycbc/neutron_stars/eos_utils.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+        Name of the equation of state.
+
+    Returns
+    -------
+    ns_compactness : float
+        Compactness parameter of the neutron star.
+    ns_b_mass : float
+        Baryonic mass of the neutron star.
+    """
+    if isinstance(ns_mass, numpy.ndarray):
```
Comment (titodalcanton) [inline pycbc/neutron_stars/eos_utils.py:132]: Sorry, I actually introduced a genuine error here:
```suggestion
    if isinstance(ns_mass, np.ndarray):
```

Comment (titodalcanton) [reply pycbc/neutron_stars/eos_utils.py:132]: Well, this is trivial enough that I can probably approve my own suggestion.

File: pycbc/distributions/constraints.py [hunk 12]
```diff
@@ -37,7 +38,12 @@ def __init__(self, constraint_arg, static_args=None, transforms=None,
                 static_args.items(), key=lambda x: len(x[0]), reverse=True))
             )
         for arg, val in static_args.items():
-            constraint_arg = constraint_arg.replace(arg, str(val))
+            swp = f"'{val}'" if isinstance(val, str) else str(val)
+            # Substitute static arg name for value if it appears in the
+            # constraint_arg string at the beginning of a word and is not
+            # followed by an underscore or equals sign
+            constraint_arg = re.sub(
+                r'\b{}(?!\_|\=)'.format(arg), swp, constraint_arg)
```
Comment (pannarale) [inline pycbc/distributions/constraints.py:46]: @SamuelH-97, can you remind us about why this change was introduced?

@cdcapano, do you see this potentially breaking `constraints.py` applications?

**Review by pannarale — COMMENTED** [review]
I only have 1 question to raise for @SamuelH-97.  In turn it raises a check for @cdcapano .

File: pycbc/distributions/constraints.py (hunk 12 above)
Comment (SamuelH-97) [reply pycbc/distributions/constraints.py:46]: I'm afraid this change was made by @a-r-williamson and we never went over the reason for it so I'm not quite certain

**Comment by titodalcanton** [comment]
@spxiwh recommends moving the large pickle file to the pycbc-config repository. I agree the file is a bit large to keep here. Will look into that.

File: pycbc/distributions/constraints.py (hunk 12 above)
Comment (pannarale) [reply pycbc/distributions/constraints.py:46]: @titodalcanton and I looked through some old tests of Andrew's and we understood that this regex call handles things like this:

"""
[static_params]
eos = 2H

[constraint-3]
name = custom
constraint_arg = remnant_mass_from_mass1_mass2_spherical_spin_eos(mass1, mass2, spin1_a, spin1_polar, eos=eos, solve=True) > 0
"""

This ensures that each injection in the injection file has a record of what eos was used (2H) and that the call to `remnant_mass_from_mass1_mass2_spherical_spin_eos` uses the keyword `eos` with value `2H` rather than `eos`.

File: pycbc/distributions/constraints.py [hunk 13]
```diff
@@ -37,7 +38,12 @@ def __init__(self, constraint_arg, static_args=None, transforms=None,
                 static_args.items(), key=lambda x: len(x[0]), reverse=True))
             )
         for arg, val in static_args.items():
-            constraint_arg = constraint_arg.replace(arg, str(val))
+            swp = f"'{val}'" if isinstance(val, str) else str(val)
+            # Substitute static arg name for value if it appears in the
+            # constraint_arg string at the beginning of a word and is not
+            # followed by an underscore or equals sign
```
Comment (pannarale) [inline pycbc/distributions/constraints.py:44]: @SamuelH-97 please add the following lines of comment:
```suggestion
            # Substitute static arg name for value if it appears in the
            # constraint_arg string at the beginning of a word and is not
            # followed by an underscore or equals sign.
            # This ensures that static_args that are also kwargs in function calls are
            # handled correctly, i.e., the kwarg is not touched while its value is replaced
            # with the static_arg value.
```

**Comment by titodalcanton** [comment]
Some notes for posterity.

Francesco and I spent some time looking into why the ISSO interpolant is needed, since it adds quite a bit of complexity. We reproduced an earlier test from Andrew, which indeed indicates that generating the injections with the exact calculation takes ~10 min, versus just a few seconds using the interpolant. This, however, seems to contradict my own timing of the ISSO solver, which indicates that the exact calculation takes just about half a millisecond for a single point, and I struggle to believe the interpolant would be significantly faster than that.

Bottom line: I propose we modify this PR to only include the exact ISSO calculation, and leave adding the interpolant as a smaller PR later, if the extra time is found to be really unacceptable.

**Comment by titodalcanton** [comment]
I have been poking at the ISSO code a bit more, so here are a few more observations.

First, I found that `PG_ISSO_solver()` is already quite fast, however the `RectBivariateSpline` is faster by orders of magnitude. Since `pycbc_create_injections` seems to invoke the constraint function millions and millions of times, using the spline does indeed reduce the execution time from several minutes (or even almost one hour with 100000 injections) to just a few seconds. So using the spline is indeed beneficial, though probably not immediately essential, as I am told that PyGRB creates the injections as a Condor job anyway.

Next, I spent some time looking for a way to make the spline pickle file significantly smaller by making the interpolation grid coarser. In doing so, I realized that `PG_ISSO_solver()` does not produce a sensible result when the spin magnitude is exactly 1:
image
This problem seems to go away by removing the bracketing used in the third root finding in `PG_ISSO_solver()`, although the curve I obtain is still not entirely smooth. I will not request that this issue is fixed in this PR, but we should fix is soon after this is merged.

Anyway, after making that change, I found that I can reduce the spline pickle file to ~15 Kb only, and still obtain an interpolation better than 1e-5 almost everywhere (note that the spline proposed in the PR does *not* satisfy the condition that the error is below 1e-5 everywhere, contrary to what the docstring of `generate_isso_bivariate_interp()` claims).

Conclusion of this secondo note: I still propose we modify this PR to only include the exact ISSO calculation (I will look into this myself next). In one or two followup PRs, we will then address the high-spin problem in `PG_ISSO_solver()` and bring in a smaller version of the interpolating spline.

File: pycbc/tmpltbank/coord_utils.py [hunk 14]
```diff
@@ -291,8 +293,15 @@ def get_random_mass(numPoints, massRangeParams):
...
-                remnant = remnant_masses(eta_nsbh, mass2_nsbh, ns_sequence,
-                                         spin1z_nsbh, 0.)
+                remnant = conversions.remnant_mass_from_mass1_mass2_cartesian_spin_eos(
+                    mass1_nsbh,
+                    mass2_nsbh,
+                    spin1x=0.0,
+                    spin1y=0.0,
+                    spin1z=spin1z_nsbh,
+                    eos=eos,
+                    interp=True
```
Comment (titodalcanton) [inline pycbc/tmpltbank/coord_utils.py:303]: I forgot this one:
```suggestion
                    eos=eos
```

File: pycbc/distributions/constraints.py (hunk 12 above)
Comment (titodalcanton) [reply pycbc/distributions/constraints.py]: Sounds good.

**Review by titodalcanton — APPROVED** [review]
There are issues with this code as noted in the previous comments, but no show stoppers at this point, so let's merge and fix those later.
